---
title: "calculate_hw_and_eigenvectors_wecare_only"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 27Aug2017

# Summary

Calculates AFs and HWE for 480 cases (before calculating eigenvectors)  
Removes 684 variants violating HWE (p <10-4 : threshold recommended by EZ)  
Some of these could be valid multiallelic variants - not verifyed at this occasion  

Eigenvectors are calculated using **44,508** common variants only:  
5% < AF  < 95% in each of the compared datasets (UBC and CBC)  

Requires accessory scripts f01_calculate_eigenvectors.R and f03_qqunif_plot.R  

Suggests two eigenvectors' outliers (> 6 SdtDev on 2nd EV): P5_E09 and P6_D05  
Additionally there are two outliers along the 4th EV: P2_C08 and P4_F10  

Input data: 239,642 vars x 480 cases (245 UBC and 235 CBC)  
Output data: 238,958 vars x 480 cases (245 UBC and 235 CBC)  

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())

# Libraries location
.libPaths("/home/alarionov/R/my_libs_r3.4.1/")

# Base folder
library(knitr)
base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_08.17"
opts_knit$set(root.dir = base_folder)
#setwd(base_folder)

# Required libraries
library(ggplot2)
library(plotly)
library(HardyWeinberg)

# Accessory functions
source(paste(base_folder, "scripts", "f03_qqunif_plot.R", sep="/"))
source(paste(base_folder, "scripts", "f01_calculate_eigenvectors.R", sep="/"))

# Filtering settings
hwe_th <- 0.0001 

```

# load_data

```{r load_data}

#base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_08.17"
load(paste(base_folder, "results", "r04_filter_cases_and_variants_wecare_only.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(variants.df)
colnames(variants.df)
variants.df[1:5,1:5]

dim(kgen.df)
colnames(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
colnames(exac.df)
exac.df[1:5,1:5]

# Check consistency of colnames and rownames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))

sum(rownames(genotypes.mx) != rownames(variants.df))
sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))

```

# recalculate_AFs

Used for selecting common variants for eigenvectors computation.  
Will be recalculated later after exclusion of eigenvectors outliers.  

```{r recalculate_AFs}

# Rename AF fields in the variants table
vars_colnames <- colnames(variants.df)
"ac_raw" -> vars_colnames[ vars_colnames == "AC" ]
"an_raw" -> vars_colnames[ vars_colnames == "AN" ]
"af_raw" -> vars_colnames[ vars_colnames == "AF" ]
vars_colnames -> colnames(variants.df)

# Function to calculate AN
get_allele_number.udf <- function(x){2*sum(!is.na(x))}

# --- Calculate total AFs --- #

# Calculate total ac, an and af
ac_all <- apply(genotypes.mx, 1, sum, na.rm=TRUE)
an_all <- apply(genotypes.mx, 1, get_allele_number.udf)
af_all <- ac_all/an_all

# Add new AFs to the variants table
variants.df <- cbind(variants.df, ac_all, an_all, af_all)

# --- Calculate ubc AFs --- #

# Prepare genotypes table
ubc_cases <- phenotypes.df$cc == 0 
sum(ubc_cases) # 245
ubc_genotypes.mx <- genotypes.mx[,ubc_cases]
dim(ubc_genotypes.mx)

# Calculate ubc ac, an and af
ac_ubc <- apply(ubc_genotypes.mx, 1, sum, na.rm=TRUE)
an_ubc <- apply(ubc_genotypes.mx, 1, get_allele_number.udf)
af_ubc <- ac_ubc/an_ubc

# Add updated AFs to variants.df
variants.df <- cbind(variants.df, ac_ubc, an_ubc, af_ubc)

# --- Calculate_cbc_AFs --- #

# Prepare genotypes table
cbc_cases <- phenotypes.df$cc == 1 
sum(cbc_cases) # 235
cbc_genotypes.mx <- genotypes.mx[,cbc_cases]
dim(cbc_genotypes.mx)

# Calculate cbc ac, an and af
ac_cbc <- apply(cbc_genotypes.mx, 1, sum, na.rm=TRUE)
an_cbc <- apply(cbc_genotypes.mx, 1, get_allele_number.udf)
af_cbc <- ac_cbc/an_cbc

# Add updated AFs to variants.df
variants.df <- cbind(variants.df, ac_cbc, an_cbc, af_cbc)

# Clean-up
rm(vars_colnames, get_allele_number.udf, ac_all, an_all, af_all, 
   cbc_cases, cbc_genotypes.mx, ac_cbc, an_cbc, af_cbc,
   ubc_cases, ubc_genotypes.mx, ac_ubc, an_ubc, af_ubc)

```

# filter_by_hwe

Using library HardyWeinberg  

```{r filter_by_hwe}

# Prepare genotypes counts
genotypes_counts <- MakeCounts(t(genotypes.mx),coding=c(0,1,2))
dim(genotypes_counts)
genotypes_counts[1:10,]

# Calculate HW p-values
hwe <- HWExactStats(genotypes_counts[,1:3], verbose=FALSE)
hwe[1:10]
variants.df <- cbind(variants.df, hwe)

# Select common variants (for QQ plot only)
common_variants <- variants.df$af_all > 0.05 & variants.df$af_all < 0.95
sum(common_variants) # 46,795

# Bonferroni-style threshold - too relaxed (EZ)
1/length(hwe) # ~4e-6
hwe_violators <- hwe < 1/length(hwe) 
sum(hwe_violators) # 605

# A stronger conventional threshold (10-4, EZ recommended between 5e-4 5e-5)
hwe_violators <- hwe < hwe_th
sum(hwe_violators) # 684

# QQ-plots for HWE
qqunif.plot(hwe, 
            main="QQ plot for all HWE p-values")

qqunif.plot(hwe[!hwe_violators], 
            main=paste("QQ plot for HWE p-values",
                "\n excluding HWE violaters (p<10-4)"))

qqunif.plot(hwe[!hwe_violators & common_variants], 
            main=paste("QQ plot for HWE p-values",
                "\n excluding HWE violaters (p<10-4) and rare variants (MAF<5%)"))

# Remove variants violating HWE 
variants.df <- variants.df[!hwe_violators,]
genotypes.mx <- genotypes.mx[!hwe_violators,]
kgen.df <- kgen.df[!hwe_violators,]
exac.df <- exac.df[!hwe_violators,]

# Check results
dim(variants.df)

# Clean-up
rm(genotypes_counts, hwe, hwe_violators, common_variants, qqunif.plot, hwe_th)

```

# calculate_eigenvectors

Requires source("f01_calculate_eigenvectors.R")  

Only common variants (0.05 < AF < 0.95 in both CBC and UBC) will be used for eigenvectors calculation.  

Note exclusion on both sides: low- and high- AFs:  
- Low AFs remove rare variants with common allele in reference genome  
- Hight AFs remove rare variants with common allele in reference genome  

```{r calculate_eigenvectors}

# --- Make subset of data with common variants only

cbc_common_vars <- variants.df$af_cbc > 0.05 & variants.df$af_cbc < 0.95
sum(cbc_common_vars) # 46,135

ubc_common_vars <- variants.df$af_ubc > 0.05 & variants.df$af_ubc < 0.95
sum(ubc_common_vars) # 46,076

common_overlap_vars <- cbc_common_vars & ubc_common_vars
sum(common_overlap_vars) # 44,508

min(variants.df$af_all[common_overlap_vars]) # ~0.05
max(variants.df$af_all[common_overlap_vars]) # ~0.95

common_overlap_genotypes.mx <- genotypes.mx[common_overlap_vars,]
dim(common_overlap_genotypes.mx)
common_overlap_genotypes.mx[1:5,1:5]

# --- Calculate eigenvectors --- #

wecare.eigen <- normalise_and_calculate_eigenvectors.udf(common_overlap_genotypes.mx)

# Clean-up
rm(cbc_common_vars, ubc_common_vars, common_overlap_vars, 
   common_overlap_genotypes.mx, normalise_and_calculate_eigenvectors.udf)

```

# plot_eigenvectors

Note manually coded varaintsxsamples values in figure titles:  
need to be corrected manually, if changed  

```{r plot_eigenvectors}

# --- Prepare data for plotting --- #

wecare.eigenvectors.df <- as.data.frame(wecare.eigen$vectors) # eigenvectors in columns

# Prepare colour scale
colours <- c("UBC" = "BLUE", "CBC" = "RED")
userColourScale <- scale_colour_manual(values=colours)

# Prepare cases lables
cases_labels <- as.vector(phenotypes.df$cc)
"CBC" -> cases_labels[cases_labels==1]
"UBC" -> cases_labels[cases_labels==0]

summary(as.factor(cases_labels))

# Prepare cases IDs (for labels on onteractive plot)
cases_IDs <- as.vector(phenotypes.df$wes_id)

# make the dataframe
data2plot.df <- cbind(cases_IDs, cases_labels, wecare.eigenvectors.df[,1:5])
colnames(data2plot.df) <- c("wes_id", "group", "ev1", "ev2", "ev3", "ev4", "ev5")

# --- Plot eig1 vs eig2 --- #

g <- ggplot(data2plot.df, aes(ev1, ev2)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(44,508 x 480)", x ="eigenvector1", y = "eigenvector2") +
  userColourScale

ggplotly(g)

# --- Plot eig2 vs eig3 --- #

g <- ggplot(data2plot.df, aes(ev2, ev3)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(44,508 x 480)", x ="eigenvector2", y = "eigenvector3") +
  userColourScale

ggplotly(g)

# --- Plot eig3 vs eig4 --- #

g <- ggplot(data2plot.df, aes(ev3, ev4)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(44,508 x 480)", x ="eigenvector3", y = "eigenvector4") +
  userColourScale

ggplotly(g)

# --- Plot eig4 vs eig5 --- #

g <- ggplot(data2plot.df, aes(ev4, ev5)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare common variants<br>(44,508 x 480)", x ="eigenvector4", y = "eigenvector5") +
  userColourScale

ggplotly(g)

# --- Clean-up --- #

rm(wecare.eigenvectors.df, g, data2plot.df, cases_IDs, cases_labels, colours, userColourScale)

```

# calculate_outliers

Explore 6 standard deviations in 5 top eigenvectors  

```{r calculate_outliers}

wecare.eigenvectors.mx <- wecare.eigen$vectors # eigenvectors in columns

# No outliers on 1st ev
ev1 <- wecare.eigenvectors.mx[,1]
ev1.positive_outliers <- ev1 > mean(ev1) + 6 * sd(ev1)
ev1.negative_outliers <- ev1 < mean(ev1) - 6 * sd(ev1)
sum(ev1.positive_outliers)
sum(ev1.negative_outliers)

phenotypes.df$wes_id[ev1.positive_outliers]
phenotypes.df$wes_id[ev1.negative_outliers]

# 2 outliers on 2nd ev: P5_E09 and P6_D05
ev2 <- wecare.eigenvectors.mx[,2]
ev2.positive_outliers <- ev2 > mean(ev2) + 6 * sd(ev2)
ev2.negative_outliers <- ev2 < mean(ev2) - 6 * sd(ev2)
sum(ev2.positive_outliers)
sum(ev2.negative_outliers)

phenotypes.df$wes_id[ev2.positive_outliers]
phenotypes.df$wes_id[ev2.negative_outliers]

# No outliers on 3rd ev
ev3 <- wecare.eigenvectors.mx[,3]
ev3.positive_outliers <- ev3 > mean(ev3) + 6 * sd(ev3)
ev3.negative_outliers <- ev3 < mean(ev3) - 6 * sd(ev3)
sum(ev3.positive_outliers)
sum(ev3.negative_outliers)

phenotypes.df$wes_id[ev3.positive_outliers]
phenotypes.df$wes_id[ev3.negative_outliers]

# 2 outliers on 4th ev: P2_C08 and P4_F10
ev4 <- wecare.eigenvectors.mx[,4]
ev4.positive_outliers <- ev4 > mean(ev4) + 6 * sd(ev4)
ev4.negative_outliers <- ev4 < mean(ev4) - 6 * sd(ev4)
sum(ev4.positive_outliers)
sum(ev4.negative_outliers)

phenotypes.df$wes_id[ev4.positive_outliers]
phenotypes.df$wes_id[ev4.negative_outliers]

# No outliers on 5th ev
ev5 <- wecare.eigenvectors.mx[,5]
ev5.positive_outliers <- ev5 > mean(ev5) + 6 * sd(ev5)
ev5.negative_outliers <- ev5 < mean(ev5) - 6 * sd(ev5)
sum(ev5.positive_outliers)
sum(ev5.negative_outliers)

phenotypes.df$wes_id[ev5.positive_outliers]
phenotypes.df$wes_id[ev5.negative_outliers]

# Plot eigenvalues
plot(wecare.eigen$values, main="Eigenvalues, wecare only")
wecare.eigen$values[1:10]

# Clean-up
rm(wecare.eigenvectors.mx, 
   ev1, ev1.positive_outliers, ev1.negative_outliers, 
   ev2, ev2.positive_outliers, ev2.negative_outliers, 
   ev3, ev3.positive_outliers, ev3.negative_outliers,
   ev4, ev4.positive_outliers, ev4.negative_outliers,
   ev5, ev5.positive_outliers, ev5.negative_outliers)

```

# data_summary

```{r data_summary}

ls()

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(variants.df)
colnames(variants.df)
variants.df[1:5,1:5]

dim(kgen.df)
colnames(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
colnames(exac.df)
exac.df[1:5,1:5]

str(wecare.eigen)

sum(colnames(genotypes.mx) != rownames(phenotypes.df))

sum(rownames(genotypes.mx) != rownames(variants.df))
sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))

```

# save_data

```{r save_data}

#base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_08.17"
save.image(paste(base_folder, "results", "r05_calculate_hw_and_egenvectors_wecare_only.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
