---
title: "filter_by_variant_effect_wecare_only"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 29Aug2017

# Summary

237,258 -> 17,857 variants  

8,936 genes:  
~2 variants (mean) per gene  
~28 inv-aggr-alt-counts (mean) per gene (median = 2, MACs)  

### Including Loss of function variants
"splice_acceptor_variant", "splice_donor_variant", "stop_gain", "stop_lost",  
"start_lost", "frameshift_variant" : 3,948 variants  

### Including dd-missense variants, likely affecting protein function
"deleterious" and "probably_damaging": 13,752 variants (of 86,342 total missenses)   
add NMD-transcripts ?  

### Including ClinSig variants
"likely_pathogenic", "risk_factor", "pathogenic", "association", "protective",  
"drug_response": 1,091 variants

### Excluding ClinSig variants
"benign", "likely_benign": 11,709 variants  

### NOT excluding multiallelic variaints  
394 multiallelic varaints in the selected  

## Invert genotypes after filtering  
SKAT-style inversion: for variants with **total** AF > 0.5  
The inverted data will be used for exploration in this script and  
for filtering in the next script  

## Explore the selected variants  
AFs, numbers of variants, inv-aggr-alt-counts (MACs) per gene etc  

Input data: 237,258 vars x 478 cases (245 UBC and 233 CBC)  
Output data: 17,857 vars x 478 cases (245 UBC and 233 CBC)  

# start_section

```{r start_section}

# Start time
Sys.time()

# Clenan-up
rm(list=ls())

# Libraries location
.libPaths("/home/alarionov/R/my_libs_r3.4.1/")

# Base folder
library(knitr)
base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_08.17"
opts_knit$set(root.dir = base_folder)
#setwd(base_folder)

# Load libraries
library(dplyr)
library(reshape2) # for melt()
library(ggplot2)
library(VennDiagram)

```

# load_data

```{r load_data}

#base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_08.17"
load(paste(base_folder, "results", "r06_exclude_eigenvectors_outliers_wecare_only.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(variants.df)
colnames(variants.df)
variants.df[1:5,1:5]

dim(kgen.df)
colnames(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
colnames(exac.df)
exac.df[1:5,1:5]

sum(rownames(genotypes.mx) != rownames(variants.df))
sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))

sum(colnames(genotypes.mx) != rownames(phenotypes.df))

```

# --- explore_variants_annotations --- #

## VEP_consequences

A combination of the below consequencies can be assigned to a variant  
"incomplete_terminal_codon_variant" added to wecare only  

```{r VEP_consequences}

sum(is.na(variants.df$Consequence)) # 0
a <- strsplit(as.vector(variants.df$Consequence),"&")
b <- unique(as.vector(unlist(a)))
b
rm(a,b)

```

## sift, polyphen

### Annotations seen in wecare-nfe

*SIFT:* "tolerated", "deleterious", "tolerated_low_confidence", "deleterious_low_confidence"  

*PolyPhen:* "benign", "possibly_damaging", "probably_damaging", "unknown"  

### Interpreting SIFT and PolyPhen scores (not used here)
*SIFT:* the higher score the less deleterious  
*PolyPhen:* the higher score the more damaging

```{r explore_SIFT_PolyPhen}

# SIFT
unique(as.vector(variants.df$SIFT_call))

# PolyPhen
unique(as.vector(variants.df$PolyPhen_call))

```

## clin_sig

### Annotations seen in wecare-nfe

*CLIN_SIG:* "likely_benign", "benign", "likely_pathogenic", "risk_factor", "not_provided", 
"uncertain_significance", "pathogenic", "association", "protective", "other", "drug_response"  

The same as in wecare-only. I have also seen "confers_sensitivity" in CLIN_SIG for gastric cases

```{r clin_sig}

unique(unlist(strsplit(as.vector(variants.df$CLIN_SIG), split="\\&")))

```

## count_multiallelic_varaints

4,754 multiallelic varaints, ~2% of all variants  
They will be preserved, as they already were split to sparate lines  
and gross HWE violators were removed  

```{r count_multiallelic_varaints}

summary(variants.df$Multiallelic) # 5,370

```

# --- Make indices --- #

## make_lof_index

```{r make_lof_index}

# Make sure there is no NAs in consequence
sum(is.na(variants.df$Consequence)) # 0

# --- Splice acceptor variants (sav) --- #

splice_acceptor_variant <- grepl( "splice_acceptor_variant", variants.df$Consequence)
sum(splice_acceptor_variant) # 343

# mean count of alleles per case in wecare
sav_wecare.mx <- genotypes.mx[splice_acceptor_variant,]
sav_wecare.counts <- apply(sav_wecare.mx, 2, sum, na.rm=TRUE)
sav_wecare.mean <- mean(sav_wecare.counts, na.rm=TRUE) # ~24
rm(sav_wecare.mx, sav_wecare.counts)

# --- Splice donor variants (sdv) --- #

splice_donor_variant <- grepl( "splice_donor_variant", variants.df$Consequence)
sum(splice_donor_variant) # 495

# mean count of alleles per case in wecare
sdv_wecare.mx <- genotypes.mx[splice_donor_variant,]
sdv_wecare.counts <- apply(sdv_wecare.mx, 2, sum, na.rm=TRUE)
sdv_wecare.mean <- mean(sdv_wecare.counts, na.rm=TRUE) # ~15
rm(sdv_wecare.mx, sdv_wecare.counts)

# --- Stop gains --- #

stop_gain <- grepl( "stop_gain", variants.df$Consequence)
sum(stop_gain) # 1,704

# mean count of alleles per case in wecare
stogv_wecare.mx <- genotypes.mx[stop_gain,]
stogv_wecare.counts <- apply(stogv_wecare.mx, 2, sum, na.rm=TRUE)
stogv_wecare.mean <- mean(stogv_wecare.counts, na.rm=TRUE) # ~55
rm(stogv_wecare.mx, stogv_wecare.counts)

# --- Stop losts --- #

# Interestingly, there are much less of stop_losts than stop_gains: 61 and 1,704 respectively. 

# A plausable explanation coulod be that stop losts can only happen on stop codons; 
# while stop gain can happen in "any" similar codon within the gene: so stop gains have much more chances to happen. 
# Another, less plausable explanation may be that stop_gains are better tolerated than stop losses. 

stop_lost <- grepl( "stop_lost", variants.df$Consequence)
sum(stop_lost) # 61

# mean count of alleles per case in wecare
stolv_wecare.mx <- genotypes.mx[stop_lost,]
stolv_wecare.counts <- apply(stolv_wecare.mx, 2, sum, na.rm=TRUE)
stolv_wecare.mean <- mean(stolv_wecare.counts, na.rm=TRUE) # ~13
rm(stolv_wecare.mx, stolv_wecare.counts)

# --- Start losts --- #

# Functional significance of start losses may be questioned: a heterozigous start loss may
# be easily compemsated by a mere expression of retained allele?? 

# However, there is not much of them... Either this is an evolutionary control or just "low base" effect (as above for the stop_losses)?

# The total number of start losses is small, so they should not influence the analysis too much anyway ... 

# Why there is no such consequence as "Start gain" ?

start_lost <- grepl( "start_lost", variants.df$Consequence)
sum(start_lost) # 85

# mean count of alleles per case in wecare
stalv_wecare.mx <- genotypes.mx[start_lost,]
stalv_wecare.counts <- apply(stalv_wecare.mx, 2, sum, na.rm=TRUE)
stalv_wecare.mean <- mean(stalv_wecare.counts, na.rm=TRUE) # ~10
rm(stalv_wecare.mx, stalv_wecare.counts)

# --- Frameshifts --- #

frameshift_variant <- grepl( "frameshift_variant", variants.df$Consequence)
sum(frameshift_variant) # 1,291

# mean count of alleles per case in wecare
fsv_wecare.mx <- genotypes.mx[frameshift_variant,]
fsv_wecare.counts <- apply(fsv_wecare.mx, 2, sum, na.rm=TRUE)
fsv_wecare.mean <- mean(fsv_wecare.counts, na.rm=TRUE) # ~99
rm(fsv_wecare.mx, fsv_wecare.counts)

# --- all LoF-s --- #

lof <- splice_acceptor_variant | splice_donor_variant | stop_gain | stop_lost | start_lost | frameshift_variant

sum(lof) # 3,948

# mean count of alleles per case in wecare
lof_wecare.mx <- genotypes.mx[lof,]
lof_wecare.counts <- apply(lof_wecare.mx, 2, sum, na.rm=TRUE)
lof_wecare.mean <- mean(lof_wecare.counts, na.rm=TRUE) # ~214
rm(lof_wecare.mx, lof_wecare.counts)

# ------- Plot average counts of lof-s per individual ------- #

# Summary table with average numbers of lof-s per individual
lof_summary_counts.df <- as.data.frame(cbind(
  lof_type=c("splice_acceptor", "splice_donior", "stop_gain", "stop_lost", "start_lost", "frameshift", "all_LoFs"),
  lof_count=c(sav_wecare.mean, sdv_wecare.mean, stogv_wecare.mean, stolv_wecare.mean, stalv_wecare.mean, fsv_wecare.mean, lof_wecare.mean)))

lof_summary_counts.df$lof_count <- as.numeric(as.vector(lof_summary_counts.df$lof_count))
lof_summary_counts.df <- lof_summary_counts.df %>% arrange(desc(lof_count))

# Order of levels defines order of bars on the plot
lof_summary_counts.df$lof_type <- factor(lof_summary_counts.df$lof_type, 
  levels=c("all_LoFs", "frameshift", "stop_gain", "splice_acceptor", "splice_donior", "stop_lost", "start_lost"))

# Plot
ggplot(lof_summary_counts.df, aes(lof_type, lof_count)) +
  geom_bar(stat="identity", fill=c("red",rep("blue",6)), alpha=0.5) +
  geom_text(aes(label=round(lof_count,1)), vjust=-1, size=5)+
  labs(x = "", y = "") + 
  ylim(0,250) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of LoF variants per individual (wecare-only)")
  
# Clean-up
rm(splice_acceptor_variant, splice_donor_variant, stop_gain, frameshift_variant, stop_lost,
   start_lost, sav_wecare.mean, sdv_wecare.mean, stogv_wecare.mean, stolv_wecare.mean,
   stalv_wecare.mean, fsv_wecare.mean, lof_summary_counts.df)

```

## make_missense_index

Selecting missense variants which are both: [ deleterious by SIFT ] AND [ probably_damaging by PolyPhen ]  

```{r make_missense_index}

missense_all <- grepl( "missense_variant", variants.df$Consequence)
sum(missense_all) # 86,342

summary(variants.df$SIFT_call) 
# note many NAs; these are when SIFT is irrelevant (e.g. non-coding)?
deleterious <- variants.df$SIFT_call == "deleterious"
summary(deleterious)
FALSE -> deleterious[is.na(deleterious)]
summary(deleterious) # 29,196

summary(variants.df$PolyPhen_call) 
# note many NAs; these are when PolyPhen is irrelevant (e.g. non-coding)?
probably_damaging <- variants.df$PolyPhen_call == "probably_damaging"
summary(probably_damaging)
FALSE -> probably_damaging[is.na(probably_damaging)]
summary(probably_damaging) # 18,918

missense <- deleterious & probably_damaging
sum(missense) # 13,752

# Plot venn diagram of selected variants by types
grid.newpage()
draw.pairwise.venn(
  area1=sum(deleterious), 
  area2=sum(probably_damaging), 
  cross.area=sum(missense), 
  category=c(
    paste("SIFT\ndeleterious\n", sum(deleterious)),
    paste("PolyPhen\nprobably_damaging\n", sum(probably_damaging))),
  fill=c("red", "blue"), 
  alpha = c(0.3, 0.3),
  cex=2, cat.fontface=4, fontfamily=3)

# mean count of missenses per case in wecare
missense_wecare.mx <- genotypes.mx[missense,]
missense_wecare.counts <- apply(missense_wecare.mx, 2, sum, na.rm=TRUE)
missense_wecare.mean <- mean(missense_wecare.counts, na.rm=TRUE)
missense_wecare.mean # ~374

# Clean-up
rm(missense_all, deleterious, probably_damaging, missense_wecare.mx, missense_wecare.counts)

```

## make_clinsig_index

ClinSig version as reported by Ensembl v87-grch37_vep_cache (Jan2017, VEP script)  
ClinVar may become a better variant-annotation database, which may be used in future.  

Selected clinsig variants  
pathogenic/likely_pathogenic, risk_factor, association, protective, drug_response  

Excluded clinsig variants  
likely_benign/benign  

All *CLIN_SIG* annotations found in the file:  
likely_benign/benign, pathogenic/likely_pathogenic,  
risk_factor, association, protective, drug_response,  
not_provided, uncertain_significance, other  

Note that sums would not need na.rm=TRUE  

```{r make_clinsig_index}

# --- likely_pathogenic and pathogenic (pat)  --- #

pathogenic_variant <- grepl( "pathogenic", variants.df$CLIN_SIG)
summary(pathogenic_variant) # 851

# mean count of alleles per case 
pat.mx <- genotypes.mx[pathogenic_variant,]
pat.counts <- apply(pat.mx, 2, sum, na.rm=TRUE)
hist(pat.counts, main="Likely_pathogenic and pathogenic variants")
pat.mean <- mean(pat.counts, na.rm=TRUE)
pat.mean # ~46
rm(pat.mx, pat.counts)

# --- risk_factor (rsf)  --- #

risk_factor_variant <- grepl( "risk_factor", variants.df$CLIN_SIG)
summary(risk_factor_variant) # 179

# mean count of alleles per case 
rsf.mx <- genotypes.mx[risk_factor_variant,]
rsf.counts <- apply(rsf.mx, 2, sum, na.rm=TRUE)
hist(rsf.counts, main="Risk factor variants")
rsf.mean <- mean(rsf.counts, na.rm=TRUE)
rsf.mean # ~65
rm(rsf.mx, rsf.counts)

# --- association (asc) --- #

association_variant <- grepl( "association", variants.df$CLIN_SIG)
summary(association_variant) # 28

# mean count of alleles per case 
asc.mx <- genotypes.mx[association_variant,]
asc.counts <- apply(asc.mx, 2, sum, na.rm=TRUE)
hist(asc.counts, main="Association variants")
asc.mean <- mean(asc.counts, na.rm=TRUE)
asc.mean # ~13
rm(asc.mx, asc.counts)

# --- protective (prt) --- #

protective_variant <- grepl( "protective", variants.df$CLIN_SIG)
summary(protective_variant) # 31

# mean count of alleles per case 
prt.mx <- genotypes.mx[protective_variant,]
prt.counts <- apply(prt.mx, 2, sum, na.rm=TRUE)
hist(prt.counts, main = "Protective variants")
prt.mean <- mean(prt.counts, na.rm=TRUE)
prt.mean # ~11
rm(prt.mx, prt.counts)

# --- drug_response (drs) --- #

drug_response_variant <- grepl( "drug_response", variants.df$CLIN_SIG)
summary(drug_response_variant) # 58

# mean count of alleles per case 
drs.mx <- genotypes.mx[drug_response_variant,]
drs.counts <- apply(drs.mx, 2, sum, na.rm=TRUE)
hist(drs.counts, main="Drug response variants")
drs.mean <- mean(drs.counts, na.rm=TRUE)
drs.mean # ~32
rm(drs.mx, drs.counts)

# --- Make toal index of selected clinsig --- #
selected_clinsig <- pathogenic_variant | risk_factor_variant | association_variant | protective_variant | drug_response_variant
summary(selected_clinsig) # 1,091

# mean count of alleles per case 
sel_clinsig.mx <- genotypes.mx[selected_clinsig,]
sel_clinsig.counts <- apply(sel_clinsig.mx, 2, sum, na.rm=TRUE)
hist(sel_clinsig.counts, main="Selected ClinSig variants")
sel_clinsig.mean <- mean(sel_clinsig.counts, na.rm=TRUE)
sel_clinsig.mean # ~145
rm(sel_clinsig.mx, sel_clinsig.counts)

# --- Make toal index of excluded clinsig --- #

excluded_clinsig <- grepl( "benign", variants.df$CLIN_SIG)
summary(excluded_clinsig) # 11,709

# mean count of alleles per case 
excl_clinsig.mx <- genotypes.mx[excluded_clinsig,]
excl_clinsig.counts <- apply(excl_clinsig.mx, 2, sum, na.rm=TRUE)
hist(excl_clinsig.counts, main="Excluded ClinSig variants")
excl_clinsig.mean <- mean(excl_clinsig.counts, na.rm=TRUE)
excl_clinsig.mean # ~3,493
rm(excl_clinsig.mx, excl_clinsig.counts)

# ------- Plot average counts of clinsig variants per individual ------- #

# Summary table with average numbers of clinsig variants per individual
clinsig_counts.df <- as.data.frame(cbind(
  clinsig_type=c("pathogenic", "risk_factor", "association", "protective", "drug_response", "All_selected_clinsig"),
  clinsig_count=c(pat.mean, rsf.mean, asc.mean, prt.mean, drs.mean, sel_clinsig.mean)))

# Convert to vector and sort
clinsig_counts.df$clinsig_count <- as.numeric(as.vector(clinsig_counts.df$clinsig_count))
clinsig_counts.df <- clinsig_counts.df %>% arrange(desc(clinsig_count))

# Order of levels defines order of bars on the plot
clinsig_counts.df$clinsig_type <- factor(clinsig_counts.df$clinsig_type, 
  levels=c("All_selected_clinsig", "risk_factor", "pathogenic", "drug_response", "association", "protective"))

# Plot
ggplot(clinsig_counts.df, aes(clinsig_type, clinsig_count)) +
  geom_bar(stat="identity", fill=c("red",rep("blue",5)), alpha=0.5) +
  geom_text(aes(label=round(clinsig_count,1)), vjust=-1, size=5) +
  labs(x = "", y = "") + 
  ylim(0,175) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of selected clinsig variants per individual (wecare-nfe)")

# --- Count inconsistent clinsig annotations --- #
sum(selected_clinsig & excluded_clinsig)

# Clean-up
rm(pathogenic_variant, risk_factor_variant, association_variant, protective_variant, drug_response_variant, 
  clinsig_counts.df, pat.mean, rsf.mean, asc.mean, prt.mean, drs.mean)

```

## explore_selected_types_per_individual_and_per_variantset

```{r explore_selected_types_per_individual_and_per_variantset}

# Count combinations of selectred categories
lof_missense <- lof & missense
clinsig_missense <- selected_clinsig & missense
lof_clinsig <- lof & selected_clinsig
lof_clinsig_missense <- lof_clinsig & missense

summary(lof_missense) # 8
summary(clinsig_missense) # 278
summary(lof_clinsig) # 115
summary(lof_clinsig_missense) # 0

selected_variants <- lof | selected_clinsig | missense

summary(selected_variants) # 18,390

# Exclude benign clinsigs

summary(selected_variants & excluded_clinsig) # 533
selected_variants <- selected_variants & ! excluded_clinsig
summary(selected_variants) # 17,857

# --- Plot mean counts of variant types per individual --- #

# Summary table with average numbers of protein-affecting per individual
summary_counts.df <- as.data.frame(cbind(
  type=c("Missense", "LoF", "ClinSig"),
  count=c(missense_wecare.mean, lof_wecare.mean, sel_clinsig.mean)))

# Convert numbers to numeric vector
summary_counts.df$count <- as.numeric(as.vector(summary_counts.df$count))

# Order of levels defines order of bars on the plot
summary_counts.df$type <- factor(summary_counts.df$type, 
  levels=c("Missense", "LoF", "ClinSig"))

# Plot bar diagramm of avg. count of selected types per individual
ggplot(summary_counts.df, aes(type, count)) +
  geom_bar(stat="identity", fill="blue", alpha=0.5) + 
  geom_text(aes(label=round(count,1)), vjust=-1, size=5)+
  ylim(0, 420) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of selected variants per individual (wecare only)")

# Plot venn diagram of selected variants by types
grid.newpage()
draw.triple.venn(
  area1=sum(lof), 
  area2=sum(selected_clinsig), 
  area3=sum(missense), 
  n12=sum(lof_clinsig), 
  n23=sum(clinsig_missense), 
  n13=sum(lof_missense),
  n123=sum(lof_clinsig_missense),
  category=c(
    paste("lof\n", sum(lof)),
    paste("clinsig\n", sum(selected_clinsig)),
    paste("missense\n", sum(missense))),
  fill = c("red", "green", "blue"), 
  alpha = c(0.3, 0.3, 0.3),
  cex=2, cat.fontface=4, fontfamily=3)

rm(missense_wecare.mean, lof_wecare.mean, sel_clinsig.mean, excl_clinsig.mean, lof, selected_clinsig, 
   excluded_clinsig, missense, clinsig_missense, lof_clinsig, lof_missense, lof_clinsig_missense, 
   summary_counts.df)

```

# keep_selected_variants_only

237,258 -> 17,857  

```{r keep_selected_variants_only}

sum(selected_variants) # 17,857

variants.df <- variants.df[selected_variants,]
kgen.df <- kgen.df[selected_variants,]
exac.df <- exac.df[selected_variants,]
genotypes.mx <- genotypes.mx[selected_variants,]


dim(variants.df)
dim(kgen.df)
dim(exac.df)
dim(genotypes.mx)

# Clean-up
rm(selected_variants)
  
```

# --- further explore selected variants --- #

This is done here for interim exploration only.  
It will be re-calculated after the variants filtering (AF < 0.05)  

## count_selected_multiallelic_varaints

394 multiallelics in selected varaints, preserved.  
May violate some implicit intuitive statistical assumptions;  
May not have major effect because of the small number (<2%) and exclusion of HWE outliers.  

```{r count_selected_multiallelic_varaints}

summary(variants.df$Multiallelic) # 394

```

## explore_AFs_in_selected_variants

Most of selected variants are rare (median ~0.1%)  
A small number of retained common variants (AF>5%) will be excluded later,  
also the weihgting will further prioritise rare variants amongst the remaining ones.  

```{r explore_AFs_in_selected_variants}

mean(variants.df$af_all) # ~1.8%
median(variants.df$af_all) # ~0.1%

hist(variants.df$af_all, main="AFs of selected variants", labels = TRUE, ylim=c(0,20000))
hist(variants.df[variants.df$af_all<0.05,"af_all"],
     ylim=c(0,20000), labels = TRUE, main="zoom to <5% (16,809 out of 17,857)", breaks=5)

```

## explore_variants_counts_per_gene

```{r explore_variants_counts_per_gene}

# Get number of genes
length(unique(variants.df$SYMBOL)) # 8,936

# Look at the top genes with high counts of variants
x <- sort(table(as.vector(variants.df$SYMBOL)), decreasing = TRUE)
x[1:200]

# Table for histogram of numbers of protein-affecting variants per gene 
# (not sure whether group_by preserves the order of genes etc, 
# but this is not important here)
y <- variants.df %>% group_by(SYMBOL) %>% summarise(n())
"vars_count" -> colnames(y)[2]

# Plot the histogram
hist(y$vars_count, main="Histogram for numbers of protein-affecting variants per gene (wecare-only)", xlab="Variants per gene", ylab="Number of genes", ylim=c(0,8000), breaks=25, labels = TRUE)

# Mean number of variants per gene
mean(y$vars_count) # ~2.0

# Clean-up
rm(x,y)

```

## invert_and_explore_inv_aggr_alt_counts_per_gene

For variants with **total** AF>0.5  MACs are calculated after genotypes inversion.  
The inversion is done here for exploration and crude counts only.  
The inversion of genotypes for analysis is done within SKAT later.  

```{r invert_and_explore_inv_aggr_alt_counts_per_gene}

# Get list of genes
genes <- unique(as.character(as.vector(variants.df$SYMBOL)))
length(genes)

# Make empty matrix for the inverted gene-aggregated data
genes_inv_aggr_data.mx <- matrix(ncol=ncol(genotypes.mx),nrow=0)
colnames(genes_inv_aggr_data.mx) <- colnames(genotypes.mx)

# Make matrix for the inverted gene-aggregated variants' counts
genes_inv_aggr_info.mx <- matrix(ncol=2, nrow=length(genes))
genes -> genes_inv_aggr_info.mx[,1]
genes -> rownames(genes_inv_aggr_info.mx)
colnames(genes_inv_aggr_info.mx) <- c("gene", "num_vars")

# Invert genotypes with total AF > 0.5
genotypes_inv.mx <- genotypes.mx
for(var in 1:nrow(variants.df)){
  
  if(variants.df[var,"af_all"] > 0.5){
    genotype <- genotypes.mx[var,]
    inv_genotype <- genotype
    0 -> inv_genotype[genotype == 2]
    2 -> inv_genotype[genotype == 0]
    genotypes_inv.mx[var,] <- inv_genotype
    rm(genotype, inv_genotype)
  }
}

# Calculate inverted gene-aggregated variants' counts
for(gene in genes){
  
  # Get the gene's variants
  vars <- variants.df$SYMBOL == gene
  
  # Record the number of variants
  num_vars <- sum(vars)
  num_vars -> genes_inv_aggr_info.mx[gene,"num_vars"]
  
  # For single-variant genes
  if(num_vars==1){
    
    # Nothing to aggregate
    counts <- genotypes_inv.mx[vars,] # vector, preserves NA
    0 -> counts[is.na(counts)] # equivalent to na.rm=TRUE below
    
  # For genes with multiple varians
  }else{
    
    # Sum alleles in the variants 
    counts <- apply(genotypes_inv.mx[vars,], 2, sum, na.rm=TRUE)
    
  }
  
  # Add result to output table
  genes_inv_aggr_data.mx <- rbind(genes_inv_aggr_data.mx, counts)
  
  # Clean-up
  rm(vars, num_vars, counts)
  
}

# Check result
# (atrix with aggregated inverted counts per gene per case)
dim(genes_inv_aggr_data.mx)
genes -> rownames(genes_inv_aggr_data.mx)
genes_inv_aggr_data.mx[1:5,1:5]

# Aggregated counts per gene
aggr_inv_AC <- apply(genes_inv_aggr_data.mx, 1, sum,na.rm=TRUE)

# Data frame with num_var-s and aggregated MACs for all genes
genes_inv_aggr_info.df <- as.data.frame(cbind(genes_inv_aggr_info.mx, aggr_inv_AC))
genes_inv_aggr_info.df$gene <- as.vector(genes_inv_aggr_info.df$gene)
genes_inv_aggr_info.df$num_vars <- as.numeric(as.vector(genes_inv_aggr_info.df$num_vars))
genes_inv_aggr_info.df$aggr_inv_AC <- as.numeric(as.vector(genes_inv_aggr_info.df$aggr_inv_AC))

# Explore result
str(genes_inv_aggr_info.df)
genes_inv_aggr_info.df[1:5,]
genes_inv_aggr_info.df[genes_inv_aggr_info.df$gene %in% c("ATM", "CHEK2"),]

# Genes with frequent protein-affecting alt alleles
x <- aggr_inv_AC[order(-aggr_inv_AC)]
mean(x) # ~28
median(x) # 2
x[1:100]

# Histograms of aggregated MACs
hist(aggr_inv_AC, labels = TRUE, ylim=c(0,10000), 
     xlab="aggregated inverted ACs (MACs) per gene", ylab="num of genes")

hist(aggr_inv_AC[aggr_inv_AC<26], breaks=c(0:25), 
     labels = TRUE, ylim=c(0,4000), main="zoom to MAC < 25 (per 478 cases)", 
     xlab="aggregated inverted ACs (MACs) per gene", ylab="num of genes")$counts

# Counts < 10 (< 1% aggregated MAF) - too rare for analysis ??
sum(aggr_inv_AC < 10) # 6,969

# Genes that may be analysed
sum(aggr_inv_AC >= 10) # 1,967 (ATM=15; CHEK2=19)

# Aggregated MACs > 100 (~ >10% aggregated MAF) - too common for being pathogenic ??
sum(aggr_inv_AC > 100) # 651
# Weighting and filtering will take care of the "common" variants in multi-variant genes

# Clean-up
rm(genes_inv_aggr_data.mx, genes_inv_aggr_info.mx, genes_inv_aggr_info.df, 
  genes, gene, aggr_inv_AC, var, x)

```

# data_summary

```{r data_summary}

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[c(1:5,591),1:5]

dim(genotypes_inv.mx)
class(genotypes_inv.mx)
genotypes_inv.mx[c(1:5,591),1:5]

dim(kgen.df)
colnames(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
colnames(exac.df)
exac.df[1:5,1:5]

dim(variants.df)
str(variants.df)
variants.df[1:5,1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))

sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))
sum(rownames(genotypes.mx) != rownames(variants.df))

```

# save_data

```{r save_data}

#base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_08.17"
save.image(paste(base_folder, "results", "r07_filter_by_variant_effect_wecare_only.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
